intake/source/discovery.py

```diff
@@ -12,19 +12,158 @@
 import time
 import logging
 
+import entrypoints
+
 from .base import DataSource
 from ..catalog.base import Catalog
+from ..config import save_conf, conf, cfile
 logger = logging.getLogger('intake')
 
 
-def autodiscover(path=None, plugin_prefix='intake_'):
-    """Scan for Intake plugin packages and return a dict of plugins.
+def autodiscover(path=None, plugin_prefix='intake_', do_package_scan=True):
+    """Discover intake drivers.
+
+    In order of decreasing precedence:
+    - Respect the 'drivers' section of the intake configuration file.
+    - Find 'intake.drivers' entrypoints provided by any Python packages in the
+      environment.
+    - Search all packages in the environment for names that begin with
+      ``intake_``. Import them and scan them for subclasses of
+      ``intake.source.base.Plugin``. This was previously the *only* mechanism
+      for auto-discoverying intake drivers, and it is maintained for backward
+      compatibility. In a future release, intake will issue a warning if any
+      packages are located by the method that do not also have entrypoints.
+
+    Parameters
+    ----------
+    path : str or None
+        Default is ``sys.path``.
+    plugin_prefix : str
+        DEPRECATED. Default is 'intake_'.
+    do_package_scan : boolean
+        Default is True.
+
+    Returns
+    -------
+    drivers : dict
+        Name mapped to driver class.
+    """
+    # Discover drivers via package scan.
+    if do_package_scan:
```

Comment:
Likely emit a `DeprecationWarning('deprecated since vX.Y.Z', stacklevel=2)` if plugin_prefix is passed. 
You can also set the default values to a sentinel (like None), to detect if people are passing the option or not.

Reply:
I would prefer to introduce the new capability without deprecating the previous behaviour, and have both live side-by-side for a while. Agree, though, that a specific warning is better than the general one.

Reply:
Roger @martindurant. We can add this later.

Reply:
it's just that the docstring says "Deprecated", and in my experience users don't read docstring. I'm just in favor of having both the docstring and the code be coherent, and give a release number so that it has a change of being cleaned up after. The `DeprecationWarning` or even `PendingDeprecationWarning` (which is foo by default), is usefull for integrator that run with `-We` to know about potential future deprecations.

Reply:
@martindurant How would you feel about a `PendingDeprecationWarning` here?

Reply:
Is that like the "future" one, that doesn't show up unless specifically requested? In that case, yes.

Reply:
Yes, I think the only warning type that is shown by default is UserWarning.

Reply:
Added and confirmed that that the `PendingDeprecationWarning` is _not_ displayed when this codepath is hit by the CLI or in interactive use in IPython.